File: python/omne_sdk/rate_limiter.py
"""
Rate limiting and backoff utilities for Omne Python SDK
"""

import asyncio
import time
from typing import List, Optional
from dataclasses import dataclass
import secrets


@dataclass
class RateLimitConfig:
    """Rate limiting configuration"""
    requests_per_second: float = 10.0
    burst_capacity: int = 100
    window_size_seconds: int = 60
    max_requests_per_window: int = 100


class AsyncRateLimiter:
    """
    Async-compatible rate limiter with token bucket and sliding window
    """
    
    def __init__(self, config: RateLimitConfig):
        self.config = config
        self.tokens = float(config.burst_capacity)
        self.last_refill = time.time()
        self.request_times: List[float] = []
        self._lock = asyncio.Lock()
    
    async def is_allowed(self) -> bool:
        """Check if request is allowed under rate limits"""
        async with self._lock:
            return self._check_limits()
    
    def _check_limits(self) -> bool:
        """Internal method to check rate limits (must hold lock)"""
        now = time.time()
        
        # Refill tokens based on elapsed time
        elapsed = now - self.last_refill
        tokens_to_add = elapsed * self.config.requests_per_second
        self.tokens = min(self.config.burst_capacity, self.tokens + tokens_to_add)
        self.last_refill = now
        
        # Clean old requests from sliding window
        cutoff_time = now - self.config.window_size_seconds
        self.request_times = [t for t in self.request_times if t > cutoff_time]
        
        # Check both token bucket and sliding window
        if (self.tokens >= 1.0 and 
            len(self.request_times) < self.config.max_requests_per_window):
            self.tokens -= 1.0
            self.request_times.append(now)
            return True
        
        return False
    
    async def wait_time(self) -> float:
        """Get estimated wait time until next request is allowed"""
        async with self._lock:
            return self._calculate_wait_time()
    
    def _calculate_wait_time(self) -> float:
        """Calculate wait time (must hold lock)"""
        now = time.time()
        
        # Time to get one token
        token_wait = 0.0
        if self.tokens < 1.0:
            tokens_needed = 1.0 - self.tokens
            token_wait = tokens_needed / self.config.requests_per_second
        
        # Time for sliding window to allow requests
        window_wait = 0.0
        cutoff_time = now - self.config.window_size_seconds
        valid_requests = [t for t in self.request_times if t > cutoff_time]
        
        if len(valid_requests) >= self.config.max_requests_per_window:
            oldest_request = min(valid_requests)
            window_wait = max(0, self.config.window_size_seconds - (now - oldest_request))
        
        return max(token_wait, window_wait)
    
    async def reset(self):
        """Reset rate limiter state"""
        async with self._lock:
            self.tokens = float(self.config.burst_capacity)
            self.last_refill = time.time()
            self.request_times.clear()


@dataclass
class BackoffConfig:
    """Exponential backoff configuration"""
    initial_delay: float = 0.1  # 100ms
    max_delay: float = 30.0     # 30 seconds
    multiplier: float = 2.0
    jitter: bool = True
    max_attempts: int = 5


class ExponentialBackoff:
    """
    Exponential backoff with jitter for retry logic
    """
    
    def __init__(self, config: BackoffConfig):
        self.config = config
        self.attempts = 0
    
    def next_delay(self) -> float:
        """Calculate next delay duration"""
        self.attempts += 1
        
        if self.attempts > self.config.max_attempts:
            raise Exception(f"Maximum retry attempts ({self.config.max_attempts}) exceeded")
        
        # Calculate base delay
        delay = self.config.initial_delay * (self.config.multiplier ** (self.attempts - 1))
        delay = min(delay, self.config.max_delay)
        
        # Add jitter if enabled
        if self.config.jitter:
            jitter_range = delay * 0.25  # 25% jitter
            jitter = secrets.randbelow(int(jitter_range * 1000)) / 1000.0
            delay += jitter
        
        return delay
    
    def reset(self):
        """Reset attempt counter"""
        self.attempts = 0
    
    @property
    def should_retry(self) -> bool:
        """Check if more retries are allowed"""
        return self.attempts < self.config.max_attempts


async def with_rate_limiting(
    rate_limiter: AsyncRateLimiter,
    backoff: Optional[ExponentialBackoff] = None,
    operation_name: str = "request"
) -> None:
    """
    Context manager for rate-limited operations
    
    Args:
        rate_limiter: Rate limiter instance
        backoff: Optional backoff for retries
        operation_name: Name for error messages
    """
    max_wait_time = 60.0  # Maximum time to wait for rate limit
    
    while True:
        if await rate_limiter.is_allowed():
            return
        
        wait_time = await rate_limiter.wait_time()
        if wait_time > max_wait_time:
            raise Exception(f"Rate limit wait time ({wait_time:.2f}s) exceeds maximum ({max_wait_time}s) for {operation_name}")
        
        if backoff:
            try:
                backoff_delay = backoff.next_delay()
                actual_delay = min(wait_time, backoff_delay)
            except Exception:
                raise Exception(f"Rate limiting failed for {operation_name} after maximum retries")
        else:
            actual_delay = wait_time
        
        await asyncio.sleep(actual_delay)


def default_rate_limiter() -> AsyncRateLimiter:
    """Create rate limiter with default configuration"""
    return AsyncRateLimiter(RateLimitConfig(
        requests_per_second=10.0,
        burst_capacity=50,
        window_size_seconds=60,
        max_requests_per_window=100
    ))


def default_backoff() -> ExponentialBackoff:
    """Create exponential backoff with default configuration"""
    return ExponentialBackoff(BackoffConfig(
        initial_delay=0.1,
        max_delay=30.0,
        multiplier=2.0,
        jitter=True,
        max_attempts=5
    ))

File: go/rate_limiter.go
package omne

import (
	"sync"
	"time"
)

// RateLimiter implements a token bucket rate limiter
type RateLimiter struct {
	mu             sync.Mutex
	tokens         float64
	capacity       float64
	refillRate     float64 // tokens per second
	lastRefill     time.Time
	requestTimes   []time.Time // Track request times for sliding window
	windowSize     time.Duration
	maxRequests    int
}

// NewRateLimiter creates a new rate limiter
func NewRateLimiter(capacity float64, refillRate float64, windowSize time.Duration, maxRequests int) *RateLimiter {
	return &RateLimiter{
		tokens:      capacity,
		capacity:    capacity,
		refillRate:  refillRate,
		lastRefill:  time.Now(),
		windowSize:  windowSize,
		maxRequests: maxRequests,
	}
}

// DefaultRateLimiter returns a rate limiter with sensible defaults
func DefaultRateLimiter() *RateLimiter {
	return NewRateLimiter(
		100,                 // 100 token capacity
		10,                  // 10 tokens per second refill rate
		time.Minute,         // 1 minute sliding window
		100,                 // 100 requests per minute max
	)
}

// Allow checks if a request should be allowed
func (rl *RateLimiter) Allow() bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()
	
	now := time.Now()
	
	// Refill tokens based on time elapsed
	elapsed := now.Sub(rl.lastRefill).Seconds()
	rl.tokens = min(rl.capacity, rl.tokens+elapsed*rl.refillRate)
	rl.lastRefill = now
	
	// Check sliding window limit
	cutoff := now.Add(-rl.windowSize)
	validRequests := 0
	for i := len(rl.requestTimes) - 1; i >= 0; i-- {
		if rl.requestTimes[i].After(cutoff) {
			validRequests++
		} else {
			// Remove old requests
			rl.requestTimes = rl.requestTimes[i+1:]
			break
		}
	}
	
	// Check both token bucket and sliding window
	if rl.tokens >= 1 && validRequests < rl.maxRequests {
		rl.tokens--
		rl.requestTimes = append(rl.requestTimes, now)
		return true
	}
	
	return false
}

// WaitTime returns the estimated wait time until the next request would be allowed
func (rl *RateLimiter) WaitTime() time.Duration {
	rl.mu.Lock()
	defer rl.mu.Unlock()
	
	now := time.Now()
	
	// Calculate wait time for token bucket
	tokensNeeded := 1 - rl.tokens
	tokenWait := time.Duration(0)
	if tokensNeeded > 0 {
		tokenWait = time.Duration(tokensNeeded/rl.refillRate) * time.Second
	}
	
	// Calculate wait time for sliding window
	cutoff := now.Add(-rl.windowSize)
	validRequests := 0
	oldestValidRequest := now
	
	for _, reqTime := range rl.requestTimes {
		if reqTime.After(cutoff) {
			validRequests++
			if reqTime.Before(oldestValidRequest) {
				oldestValidRequest = reqTime
			}
		}
	}
	
	windowWait := time.Duration(0)
	if validRequests >= rl.maxRequests {
		windowWait = rl.windowSize - now.Sub(oldestValidRequest)
		if windowWait < 0 {
			windowWait = 0
		}
	}
	
	// Return the maximum of the two wait times
	if tokenWait > windowWait {
		return tokenWait
	}
	return windowWait
}

// Reset resets the rate limiter state
func (rl *RateLimiter) Reset() {
	rl.mu.Lock()
	defer rl.mu.Unlock()
	
	rl.tokens = rl.capacity
	rl.lastRefill = time.Now()
	rl.requestTimes = nil
}

// ExponentialBackoff implements exponential backoff with jitter
type ExponentialBackoff struct {
	initialDelay time.Duration
	maxDelay     time.Duration
	multiplier   float64
	jitter       bool
	attempts     int
}

// NewExponentialBackoff creates a new exponential backoff instance
func NewExponentialBackoff(initialDelay, maxDelay time.Duration, multiplier float64, jitter bool) *ExponentialBackoff {
	return &ExponentialBackoff{
		initialDelay: initialDelay,
		maxDelay:     maxDelay,
		multiplier:   multiplier,
		jitter:       jitter,
	}
}

// DefaultExponentialBackoff returns backoff with sensible defaults
func DefaultExponentialBackoff() *ExponentialBackoff {
	return NewExponentialBackoff(
		100*time.Millisecond, // Start with 100ms
		30*time.Second,       // Max 30 seconds
		2.0,                  // Double each time
		true,                 // Add jitter
	)
}

// NextDelay calculates the next delay duration
func (eb *ExponentialBackoff) NextDelay() time.Duration {
	eb.attempts++
	
	delay := eb.initialDelay
	for i := 1; i < eb.attempts; i++ {
		delay = time.Duration(float64(delay) * eb.multiplier)
		if delay > eb.maxDelay {
			delay = eb.maxDelay
			break
		}
	}
	
	if eb.jitter {
		// Add up to 25% jitter
		jitterMs := int64(float64(delay.Milliseconds()) * 0.25)
		if jitterMs > 0 {
			// Use secure random for jitter
			jitterBytes, _ := SecureRandomBytes(8)
			jitterValue := int64(jitterBytes[0]) % jitterMs
			delay += time.Duration(jitterValue) * time.Millisecond
		}
	}
	
	return delay
}

// Reset resets the backoff attempts counter
func (eb *ExponentialBackoff) Reset() {
	eb.attempts = 0
}

// min helper function
func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}